src/units/abilities.cpp

```diff
@@ -1126,6 +1126,76 @@ int attack_type::modified_damage() const
 	return damage_value;
 }
 
+static void alignment_list(unit_ability_list& abil_list)
+{	//check if have abilities or specials alignment and remove alignment bad specified
+	const std::set<std::string> checking_alignment = {"neutral", "lawful", "chaotic", "liminal"};
+	for(unit_ability_list::iterator i = abil_list.begin(); i != abil_list.end();) {
+		if(checking_alignment.count((*i->ability_cfg)["alignment"].str()) == 0){
+			i = abil_list.erase(i);
+		} else {
+			++i;
+		}
+	}
+}
+
+static void choose_alignment(unit_ability_list& abil_list, unit_alignments::type& alignment, bool& cumulative)
+{
+	std::map<std::string, int> values;
+	std::map<std::string, bool> cumulatives;
+	for(const auto& i : abil_list) {
+		const auto& alignment = (*i.ability_cfg)["alignment"].str();
+		values[alignment] += 1;
+		if((*i.ability_cfg)["cumulative"].to_bool()){
+			cumulatives[alignment] = true;
+		}
+	}
+	int is_liminal = values["liminal"], is_chaotic = values["chaotic"], is_lawful = values["lawful"], is_neutral = values["neutral"];
+	std::string str_alignment;
```

Comment:
It's better to use enum values, and we already have the `unit_alignments::type` enum for this:

```
	std::optional<unit_alignments::type> new_align;
	if(is_liminal > 0 && is_liminal >= is_chaotic && is_liminal >= is_lawful && is_liminal >= is_neutral){
		new_align = unit_alignments::type::liminal;
	}
	else if((is_neutral > 0 && is_neutral > is_chaotic && is_neutral > is_lawful) || (is_lawful > 0 && is_lawful == is_chaotic)){
		new_align = unit_alignments::type::neutral;
	}
	else if( is_lawful > 0 && is_lawful > is_chaotic){
		new_align = unit_alignments::type::lawful;
	}
	else if( is_chaotic > 0 && is_chaotic > is_lawful){
		new_align = unit_alignments::type::chaotic;
	}
	if(new_align) {
		alignment = *new_align;
		cumulative = cumulatives[unit_alignments::get_string(alignment)];
	}
```

Reply:
except what if i could use unit::alignments::type in if loop before use if(new_align) i would done it in begun, it why a use sstring variable in first time before to define alignment variable with.